lib/mongoid/criteria/queryable/extensions/numeric.rb

```diff
@@ -51,7 +51,7 @@ module ClassMethods
             #
             # @return [ Object ] The converted number.
             def __numeric__(object)
-              object.to_s =~ /(\A[-+]?[0-9]+\z)|(\.0+\z)|(\.\z)/ ? object.to_i : Float(object)
+              object.to_s.match?(/\A[-+]?[0-9]*\.?0*\z/) ? object.to_i : Float(object)
```

Comment:
Regexes are functionally equivalent. My new one is ~50% faster.

```ruby
Benchmark.bm do |bm|
  bm.report { x.times { '21329.2328'.match?(/(\A[-+]?[0-9]+\z)|(\.0+\z)|(\.\z)/) } }
  bm.report { x.times { '21329.2328'.match?(/\A[-+]?[0-9]*\.?0*\z/) } }
end
```

Reply:
They are not equivalent:

```
irb(main):001:0> /(\A[-+]?[0-9]+\z)|(\.0+\z)|(\.\z)/ =~ ''
=> nil
irb(main):002:0> /\A[-+]?[0-9]*\.?0*\z/ =~ ''
=> 0
```

Please make one change per PR.

Reply:
I'll split this out.

Reply:
Reverted